backend/routes/portfolios.js

```diff
@@ -0,0 +1,134 @@
+const { PrismaClient } = require("../generated/prisma");
+const prisma = new PrismaClient();
+
+const express = require("express");
+const { all } = require("./auth");
+const app = express();
+app.use(express.json());
+const router = express.Router({ mergeParams: true });
+require("dotenv").config();
+
+// make new portfolio
+
+router.post("/", async (req, res) => {
+  const userId = req.session.userId;
+  const { name, description } = req.body;
+
+  if (!userId) {
+    res.status(401).json({ message: "not logged in" });
+  }
+
+  const newPortfolio = await prisma.portfolio.create({
+    data: {
+      name,
+      description,
+      companiesIds: [],
+      userId,
+    },
+  });
+  res.status(200).json(newPortfolio);
+});
+
+router.get("/", async (req, res) => {
+  const userId = req.session.userId;
+  const allPortfolios = await prisma.portfolio.findMany({
+    where: {
+      userId,
+    },
+  });
+  if (allPortfolios == null) {
+    res.status(200).json({ message: "this user has no current portfolios" });
+  }
+  res.status(200).json(allPortfolios);
+});
+
+router.delete("/:id", async (req, res) => {
+  const portfolioId = parseInt(req.params.id);
+  const portfolio = await prisma.portfolio.delete({
+    where: {
+      id: portfolioId,
+    },
+  });
+  if (portfolio == null) {
+    res.status(400).json({ message: "failed to delete" });
+  }
+  res.json(portfolio);
+});
+
+// delete company from portfolio
+router.delete("/:id/:companyId", async (req, res) => {
+  const portfolioId = parseInt(req.params.id);
+  const companyId = parseInt(req.params.companyId);
+  const portfolio = await prisma.portfolio.findUnique({
+    where: {
+      id: portfolioId,
+    },
+  });
+  const array = portfolio.companiesIds.filter((val) => val !== companyId);
+
+  const updatedPortfolio = await prisma.portfolio.update({
+    where: { id: portfolioId },
+    data: {
+      companiesIds: array,
+    },
+  });
+
+  if (updatedPortfolio == null) {
+    res.status(400).json({ message: "failed to delete" });
+  }
+  res.json(updatedPortfolio);
+});
+
+router.get("/:id", async (req, res) => {
+  const portfolioId = parseInt(req.params.id);
+
+  const userId = req.session.userId;
+
+  if (userId == null) {
+    res.status(401).json({ message: "you are not logged in" });
+  }
```

Comment:
consider adding a [middleware](https://expressjs.com/en/guide/using-middleware.html) to enforce all endpoints should have user logged in

```
// a middleware function with no mount path. This code is executed for every request to the router
router.use((req, res, next) => {
  const userId = req.session.userId;
  if (userId == null) {
    res.status(401).json({ message: "you are not logged in" });
    return;
  }
  next()
})
```